shared/trieutil/merkle_trie.go

```diff
@@ -35,3 +40,38 @@ func PrevPowerOf2(n int) int {
 	}
 	return 2 * PrevPowerOf2(n/2)
 }
+
+// MerkleTree returns all the nodes in a merkle tree from inputting merkle leaves.
+//
+// Spec pseudocode definition:
+//   def merkle_tree(leaves: Sequence[Hash]) -> Sequence[Hash]:
+//    padded_length = get_next_power_of_two(len(leaves))
+//    o = [Hash()] * padded_length + list(leaves) + [Hash()] * (padded_length - len(leaves))
+//    for i in range(padded_length - 1, 0, -1):
+//        o[i] = hash(o[i * 2] + o[i * 2 + 1])
+//    return o
+func MerkleTree(leaves [][]byte) [][]byte {
+	paddedLength := NextPowerOf2(len(leaves))
+	parents := make([][]byte, paddedLength)
+	paddedLeaves := make([][]byte, paddedLength-len(leaves))
+
+	for i := 0; i < len(parents); i++ {
+		parents[i] = params.BeaconConfig().ZeroHash[:]
+	}
+	for i := 0; i < len(paddedLeaves); i++ {
+		paddedLeaves[i] = params.BeaconConfig().ZeroHash[:]
+	}
+
+	merkleTree := make([][]byte, 0, len(parents)+len(leaves)+len(paddedLeaves))
+	merkleTree = append(merkleTree, parents...)
```

Comment:
I wonder if copy is faster than append here? Given the lengths are all known

Reply:
Will benchmark :)

Reply:
Copy is faster. Tested with 2**20 leaves.
**Append:**
BenchmarkMerkleTree_Generate-4   	      10	 226862027 ns/op
**Copy:**
BenchmarkMerkleTree_Generate-4   	      20	  67944152 ns/op

Reply:
Updated to use copy, thanks for the great suggestion!